aim/storage/locking.py

```diff
@@ -0,0 +1,115 @@
+import os
+import logging
+
+from filelock import BaseFileLock, SoftFileLock, UnixFileLock, has_fcntl
+
+from cachetools.func import ttl_cache
+from psutil import disk_partitions
+
+from typing import Optional, Union, Dict, Set
+
+logger = logging.getLogger(__name__)
+
+
+class FileSystemInspector:
+    """
+    This class is used to inspect the file system state and determine if the
+    file being created can use `fcntl` / `flock` or should use soft file locks.
+
+    This can be achieved by finding the device owning the parent directory and
+    mapping it to the partition and finally checking the filesystem type.
+
+    We prefer using soft file locks *only* if there is a risk of the filesystem
+    not being able to handle `fcntl` implementations.
+    (e.g. remote filesystems, including NFS<4, BeeGFS, Lustre, etc.)
+    """
+    GLOBAL_FILE_LOCK_CAPABLE_FILESYSTEMS: Set[str] = {
+        # The most popular local filesystems
+        'ext3', 'ext4', 'xfs', 'zfs', 'apfs',
+        # We include only NFS v4 filesystems here (yet).
+        # NFS v3 does not support file locks.
+        'nfs4',
+    }
+
+    @classmethod
+    @ttl_cache(ttl=10)
+    def dev_fstype_mapping(cls) -> Dict[int, str]:
+        """
+        Returns a mapping of device numbers to filesystem types.
+        """
+        mapping: Dict[int, str] = {}
+        for partition in disk_partitions(all=True):
+            try:
+                device_id = os.lstat(partition.mountpoint).st_dev
+            except (OSError, ValueError, AttributeError):
+                continue
+            mapping[device_id] = partition.fstype
+        return mapping
+
+    @classmethod
+    def get_fstype(cls, path: str) -> Optional[str]:
+        """
+        Returns the filesystem type of the file / directory of the given path.
+
+        If the path does not exist, `None` is returned.
+        """
+        try:
+            stat = os.stat(path)
+        except (OSError, ValueError, AttributeError):
+            return None
+        return cls.dev_fstype_mapping().get(stat.st_dev)
+
+    @classmethod
+    def needs_soft_lock(cls, path: str) -> bool:
+        """
+        Returns `True` if the file system of the given path is not capable of
+        using `fcntl` locks.
+        """
+        if not has_fcntl:
+            # We can't use `fcntl` locks on this platform no matter the
+            # underlying filesystem type.
+            logger.warning("fcntl is not available on this platform, "
```

Comment:
@mahnerak throwing warnings each time acquiring a lock doesn't make sense, let's do it once only